### PR TITLE
E2E test for issuance

### PR DESCRIPTION
**Pull request by alexeykoren**
Tests for ZSA issuance and transfer

### CODE REVIEW

File: src/builder.rs [hunk 1]
```diff
@@ -397,14 +397,15 @@ impl Builder {
         let anchor = self.anchor;
 
         // Determine the value balance for this bundle, ensuring it is valid.
-        let value_balance = pre_actions
+        let native_value_balance = pre_actions
             .iter()
+            .filter(|a| a.output.note_type.is_native().into())
```
Comment (PaulLaux) [inline src/builder.rs:402]: Should remove `filter`

File: tests/builder.rs [hunk 2]
```diff
@@ -1,4 +1,6 @@
 use incrementalmerkletree::{bridgetree::BridgeTree, Hashable, Tree};
```
Comment (PaulLaux) [inline tests/builder.rs:1]: split in a different file for issuance tests

File: tests/builder.rs [hunk 3]
```diff
@@ -112,3 +124,77 @@ fn bundle_chain() {
...
+        // Take first note from first action
+        let note = issuing_bundle.actions().first().unwrap().notes().first();
+
+        // Use the tree with a single leaf.
+        let cmx: ExtractedNoteCommitment = note.commitment().into();
+        let leaf = MerkleHashOrchard::from_cmx(&cmx);
+        let mut tree = BridgeTree::<MerkleHashOrchard, 32>::new(0);
+        tree.append(&leaf);
+        let position = tree.witness().unwrap();
+        let root = tree.root(0).unwrap();
```
Comment (PaulLaux) [inline tests/builder.rs:175]: move right after the issuance itself.

File: tests/builder.rs [hunk 4]
```diff
@@ -112,3 +124,77 @@ fn bundle_chain() {
...
+            builder.add_recipient(None, recipient, NoteValue::from_raw(5000), note_type, None),
+            Ok(())
+        );
+        let unauthorized = builder.build(&mut rng).unwrap();
+        let sighash = unauthorized.commitment().into();
+        let proven = unauthorized.create_proof(&pk, &mut rng).unwrap();
+        proven
+            .apply_signatures(&mut rng, sighash, &[SpendAuthorizingKey::from(&sk)])
+            .unwrap()
+    };
```
Comment (PaulLaux) [inline tests/builder.rs:196]: Let's test all 5 cases:
- trivial (already implemented)
- only non-native: 1 spend - 2 recipient notes
- only non-native: 2 spend - 1 recipient notes and padding 
- only non-native: 2 spend - 2 recipient notes with different denomination.
- mix with native notes.

File: tests/zsa.rs [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+            recipient,
+            NoteValue::from_raw(1),
+            note2.note_type(),
+            None
+        ),
+        Ok(())
+    );
+    let _result: Bundle<InProgress<Unproven, orchard::builder::Unauthorized>, i64> =
+        builder.build(&mut rng).unwrap();
+}
```
Comment (PaulLaux) [inline tests/zsa.rs:700]: for panic testing inside `build` use
```rust
    let result = std::panic::catch_unwind(|| {
        let _res : Result<Bundle<_, i64>, _> = builder.build(OsRng);
    });

    assert!(result.is_err());
```

File: tests/zsa.rs [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+
+    // Take notes from first action
+    let note1 = issuing_bundle
+        .actions()
+        .first()
+        .unwrap()
+        .notes()
+        .get(0)
+        .unwrap();
+    let note2 = issuing_bundle
```
Comment (PaulLaux) [inline tests/zsa.rs:266]: Please create a getter to return all notes from all actions in `issue_bundle`.
`let notes :: Vec<&Notes> = issuing_bundle.get_all_notes();`
and replace in all positions.

We will also need this down the road for consensus tree building.

**Review by PaulLaux — COMMENTED** [review]
Overall good work, added comments required to finalize.

File: tests/builder.rs (hunk 2 above)
Comment (alexeykoren) [reply tests/builder.rs]: done

File: tests/builder.rs (hunk 3 above)
Comment (alexeykoren) [reply tests/builder.rs]: done

File: tests/builder.rs (hunk 4 above)
Comment (alexeykoren) [reply tests/builder.rs]: done

File: tests/zsa.rs [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+) -> Bundle<Authorized, i64> {
+    let unauthorized = builder.build(&mut rng).unwrap();
+    let sighash = unauthorized.commitment().into();
+    let proven = unauthorized.create_proof(&pk, &mut rng).unwrap();
+    proven
+        .apply_signatures(&mut rng, sighash, &[SpendAuthorizingKey::from(&sk)])
+        .unwrap()
+}
+
+pub fn build_merkle_tree_2_leaves(note1: &Note, note2: &Note) -> (MerklePath, MerklePath, Anchor) {
```
[the remark this answers is not in the document]
Comment (alexeykoren) [reply tests/zsa.rs:79]: done

File: tests/zsa.rs [hunk 8]
```diff
@@ -0,0 +1,700 @@
...
+        .is_ok());
+
+    let issuing_bundle = sign_issuing_bundle(unauthorized, rng, isk);
+
+    // Take first note from first action
+    let note = issuing_bundle.actions().first().unwrap().notes().first();
+
+    let (merkle_path, anchor) = build_merkle_tree(note);
+
+    verify_issuing_bundle(&issuing_bundle);
```
[the remark this answers is not in the document]
Comment (alexeykoren) [reply tests/zsa.rs:144]: done

File: tests/zsa.rs [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+        .actions()
+        .get(1)
+        .unwrap()
+        .notes()
+        .get(0)
+        .unwrap();
+
+    let (merkle_path1, merkle_path2, anchor) = build_merkle_tree_2_leaves(note1, note2);
+
+    verify_issuing_bundle(&issuing_bundle);
```
[the remark this answers is not in the document]
Comment (alexeykoren) [reply tests/zsa.rs:670]: done

File: tests/zsa.rs [hunk 10]
```diff
@@ -0,0 +1,700 @@
...
+        sign_shielded_bundle(builder, rng, pk, sk)
+    };
+
+    // Verify the shielded bundle
+    verify_bundle(&shielded_bundle, &vk);
+}
+
+/// Issue 2 ZSA notes of different asset types
+#[test]
+#[should_panic]
```
[the remark this answers is not in the document]
Comment (alexeykoren) [reply tests/zsa.rs:621]: done

File: tests/zsa.rs (hunk 6 above)
Comment (alexeykoren) [reply tests/zsa.rs:266]: done

**Review by PaulLaux — APPROVED** [review]

**Pull request merged**